valkyrie/lib/valkyrie/storage/disk.rb

```diff
@@ -33,5 +34,21 @@ def find_by(id:)
       return unless handles?(id: id)
       Valkyrie::StorageAdapter::File.new(id: Valkyrie::ID.new(id.to_s), io: ::File.open(file_path(id), 'rb'))
     end
+
+    class BucketedStorage
+      attr_reader :base_path
+      def initialize(base_path:)
+        @base_path = base_path
+      end
+
+      def generate(resource:, file:)
+        Pathname.new(base_path).join(*bucketed_path(resource.id)).join(file.original_filename)
+      end
+
+      def bucketed_path(id)
+        cleaned_id = id.to_s.delete("-")
+        cleaned_id[0..5].chars.each_slice(2).map(&:join) + [cleaned_id]
```

Comment:
In Hyrax, I've seen the same task accomplished with `id.scan(/..?/).first(4).push(id)` — maybe this is faster because it doesn't use regexes?

Reply:
Easy to find out.

Reply:
```
irb(main):013:0> Benchmark.ips do |x|
irb(main):014:1* x.report("Figgy Bucket")
irb(main):015:1> ^C
irb(main):015:0> Benchmark.ips do |x|
irb(main):016:1* x.report("Figgy Bucket") do
irb(main):017:2* cleaned_id = id.to_s.delete("-")
irb(main):018:2> cleaned_id[0..5].chars.each_slice(2).map(&:join) + [cleaned_id]
irb(main):019:2> end
irb(main):020:1> x.report("Hyrax Bucket") do
irb(main):021:2* id.scan(/..?/).first(4).push(id)
irb(main):022:2> end
irb(main):023:1> x.compare!
irb(main):024:1> end
Warming up --------------------------------------
        Figgy Bucket    18.094k i/100ms
        Hyrax Bucket    15.241k i/100ms
Calculating -------------------------------------
        Figgy Bucket    210.235k (± 4.2%) i/s -      1.049M in   5.000852s
        Hyrax Bucket    163.805k (± 4.2%) i/s -    823.014k in   5.033179s

Comparison:
        Figgy Bucket:   210234.7 i/s
        Hyrax Bucket:   163805.2 i/s - 1.28x  slower
```
It's a little faster doing the each_slice, but pretty negligible. Thoughts?

Reply:
They are both a little opaque, so going with the slightly faster version seems better to me.